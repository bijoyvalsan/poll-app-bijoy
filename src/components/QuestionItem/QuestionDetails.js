import React from "react";
import PropTypes from "prop-types";

import SaveVote from "../Button";

import {
  StyledContainer,
  StyledContainerItem,
  StyledDeailsContainer,
  StyledDeailsColumn,
  StyledDeailsRow,
  StyledSubTitle,
  StyledBox
} from "./QuestionDetails.styled";

const QuestionDetails = ({ question, onVote }) => (
  <StyledBox>
    <StyledSubTitle>Question: {question.question}</StyledSubTitle>
    <StyledContainer>
      {question.choices.map((c, i) => (
        <StyledContainerItem key={c.url}>
          <StyledDeailsContainer>
            <StyledDeailsColumn>
              <StyledDeailsRow>
                Choice {i + 1}: {c.choice}
              </StyledDeailsRow>
            </StyledDeailsColumn>
            <StyledDeailsColumn>
              <StyledDeailsRow>Votes: {c.votes}</StyledDeailsRow>
            </StyledDeailsColumn>
            <StyledDeailsColumn>
              <StyledDeailsRow>Percentage: {c.percentage}%</StyledDeailsRow>
            </StyledDeailsColumn>
            <StyledDeailsColumn>
              <SaveVote.Button
                data-testid={`save-vote-${i}`}
                onClick={onVote(c.url)}
              >
                Save Vote
              </SaveVote.Button>
            </StyledDeailsColumn>
          </StyledDeailsContainer>
        </StyledContainerItem>
      ))}
    </StyledContainer>
  </StyledBox>
);

QuestionDetails.defaultProps = {
  onVote: () => {}
};

QuestionDetails.propTypes = {
  onVote: PropTypes.func,
  question: PropTypes.shape({
    id: PropTypes.string,
    question: PropTypes.string,
    choices: PropTypes.arrayOf(
      PropTypes.shape({
        choice: PropTypes.string.isRequired,
        url: PropTypes.string.isRequired,
        votes: PropTypes.number.isRequired,
        percentage: PropTypes.number.isRequired
      })
    )
  })
};

export default QuestionDetails;
