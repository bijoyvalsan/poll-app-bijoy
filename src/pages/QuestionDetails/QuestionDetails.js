import React, { useEffect } from "react";
import { useParams } from "react-router-dom";
import PropTypes from "prop-types";
import { Question, Page } from "../../components";

const QuestionDetails = ({
  getQuestionById,
  isLoading,
  question,
  saveVoteForQuestion
}) => {
  const { questionId } = useParams();
  useEffect(() => {
    getQuestionById(decodeURIComponent(questionId));
  }, [getQuestionById, questionId]);

  const handleOnVoteClick = urlPath => {
    return () => {
      saveVoteForQuestion(urlPath);
    };
  };

  return (
    <Page title="Question Details">
      <div>
        {isLoading && <Question.QuestionDetailsSkeleton />}
        {!isLoading && (
          <>
            <Question.QuestionDetails
              question={question}
              onVote={handleOnVoteClick}
            />
          </>
        )}
      </div>
    </Page>
  );
};

QuestionDetails.defaultProps = {
  getQuestionById: () => {},
  saveVoteForQuestion: () => {}
};

QuestionDetails.propTypes = {
  getQuestionById: PropTypes.func,
  saveVoteForQuestion: PropTypes.func,
  question: PropTypes.shape({
    id: PropTypes.string,
    question: PropTypes.string,
    choices: PropTypes.arrayOf(
      PropTypes.shape({
        choice: PropTypes.string.isRequired,
        url: PropTypes.string.isRequired,
        votes: PropTypes.number.isRequired,
        percentage: PropTypes.number.isRequired
      })
    )
  }),
  isLoading: PropTypes.bool.isRequired
};

export default QuestionDetails;
