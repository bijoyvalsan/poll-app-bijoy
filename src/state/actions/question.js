import { fetchQuestionById, saveVote } from "../../services/questionServices";

export const QUESTION_SINGLE_FETCH = "QUESTION_SINGLE_FETCH";
export const QUESTION_SINGLE_FETCH_SUCCSSES = "QUESTION_SINGLE_FETCH_SUCCSSES";
export const QUESTION_SINGLE_FETCH_ERROR = "QUESTION_SINGLE_FETCH_ERROR";

export const QUESTION_SINGLE_VOTE = "QUESTION_SINGLE_VOTE";
export const QUESTION_SINGLE_VOTE_SUCCSSES = "QUESTION_SINGLE_VOTE_SUCCSSES";
export const QUESTION_SINGLE_VOTE_ERROR = "QUESTION_SINGLE_VOTE_ERROR";

export const getQuestionById = urlPath => dispatch => {
  dispatch({
    type: QUESTION_SINGLE_FETCH
  });

  return fetchQuestionById(urlPath)
    .then(payload => {
      dispatch({
        type: QUESTION_SINGLE_FETCH_SUCCSSES,
        payload
      });
    })
    .catch(() => {
      dispatch({
        type: QUESTION_SINGLE_FETCH_ERROR
      });
    });
};

export const saveVoteForQuestion = urlPath => dispatch => {
  dispatch({
    type: QUESTION_SINGLE_VOTE
  });

  return saveVote(urlPath)
    .then(payload => {
      dispatch({
        type: QUESTION_SINGLE_VOTE_SUCCSSES,
        payload
      });
    })
    .catch(() => {
      dispatch({
        type: QUESTION_SINGLE_VOTE_ERROR
      });
    });
};
