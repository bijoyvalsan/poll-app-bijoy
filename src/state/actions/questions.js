import { fetchQuestions } from "../../services/questionServices";

export const QUESTIONS_FETCH = "QUESTIONS_FETCH";
export const QUESTIONS_FETCH_SUCCSSES = "QUESTIONS_FETCH_SUCCSSES";
export const QUESTIONS_FETCH_ERROR = "QUESTIONS_FETCH_ERROR";

export const getQuestions = () => dispatch => {
  dispatch({
    type: QUESTIONS_FETCH
  });

  fetchQuestions()
    .then(payload => {
      dispatch({
        type: QUESTIONS_FETCH_SUCCSSES,
        payload
      });
    })
    .catch(() => {
      dispatch({
        type: QUESTIONS_FETCH_ERROR
      });
    });
};
