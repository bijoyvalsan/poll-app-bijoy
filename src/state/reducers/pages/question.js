import {
  QUESTION_SINGLE_FETCH,
  QUESTION_SINGLE_FETCH_SUCCSSES,
  QUESTION_SINGLE_FETCH_ERROR,
  QUESTION_SINGLE_VOTE,
  QUESTION_SINGLE_VOTE_SUCCSSES,
  QUESTION_SINGLE_VOTE_ERROR
} from "../../actions/question";

export const initialState = {
  isLoading: true,
  hasError: false,
  question: {}
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case QUESTION_SINGLE_FETCH: {
      return {
        ...state,
        isLoading: true,
        hasError: false
      };
    }
    case QUESTION_SINGLE_FETCH_SUCCSSES: {
      const data = action.payload;
      return {
        ...state,
        isLoading: false,
        hasError: false,
        question: data
      };
    }
    case QUESTION_SINGLE_FETCH_ERROR: {
      return {
        ...state,
        isLoading: false,
        hasError: true
      };
    }
    case QUESTION_SINGLE_VOTE: {
      return {
        ...state,
        isLoading: true,
        hasError: false
      };
    }
    case QUESTION_SINGLE_VOTE_SUCCSSES: {
      const data = action.payload;
      return {
        ...state,
        isLoading: false,
        hasError: false,
        question: {
          ...state.question,
          choices: state.question.choices.map(c => {
            if (c.url === data.url) {
              return data;
            }
            return c;
          })
        }
      };
    }
    case QUESTION_SINGLE_VOTE_ERROR: {
      return {
        ...state,
        isLoading: false,
        hasError: true
      };
    }
    default:
      return state;
  }
};

export default reducer;
