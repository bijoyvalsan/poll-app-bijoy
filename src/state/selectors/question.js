// import { formatDate } from '../../helpers/date';

export const getQuestionDetailForListPage = state => {
  const selectedQuestion = state.pages.question.question;
  if (Object.keys(selectedQuestion).length === 0) {
    return {};
  }

  const totalVotes = selectedQuestion.choices.reduce((acc, currentChoice) => {
    const total = acc + currentChoice.votes;
    return total;
  }, 0);

  return {
    id: selectedQuestion.url,
    question: selectedQuestion.question,
    choices: selectedQuestion.choices.map(c => {
      const votePercentage = parseFloat(c.votes) / parseFloat(totalVotes);
      return {
        ...c,
        percentage: Number.isNaN(votePercentage)
          ? 0
          : Math.round(votePercentage * 100)
      };
    })
  };
};

export const isQuestionDetailPageLoading = state => {
  return state.pages.question.isLoading;
};
