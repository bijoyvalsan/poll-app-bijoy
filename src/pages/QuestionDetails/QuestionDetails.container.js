import { connect } from "react-redux";

import QuestionDetails from "./QuestionDetails";

import {
  getQuestionById,
  saveVoteForQuestion
} from "../../state/actions/question";
import {
  getQuestionDetailForListPage,
  isQuestionDetailPageLoading
} from "../../state/selectors/question";

const mapStateToProps = state => ({
  question: getQuestionDetailForListPage(state),
  isLoading: isQuestionDetailPageLoading(state)
});

const mapDispatchToProps = {
  getQuestionById,
  saveVoteForQuestion
};

export default connect(mapStateToProps, mapDispatchToProps)(QuestionDetails);
