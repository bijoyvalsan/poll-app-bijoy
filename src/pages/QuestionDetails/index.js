import QuestionDetails from "./QuestionDetails.container";

export default QuestionDetails;
