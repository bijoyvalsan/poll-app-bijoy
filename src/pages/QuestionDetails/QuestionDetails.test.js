import React from "react";
import { render, waitFor, screen, fireEvent } from "@testing-library/react";

import QuestionDetails from "./QuestionDetails";
import * as QuestionActions from "../../state/actions/question";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    questionId: "testurl"
  })
}));

describe("QuestionDetails Component", () => {
  let getQuestionByIdSpy;
  let saveVoteForQuestionSpy;

  beforeEach(() => {
    getQuestionByIdSpy = jest.spyOn(QuestionActions, "getQuestionById");
    saveVoteForQuestionSpy = jest.spyOn(QuestionActions, "saveVoteForQuestion");
  });

  afterEach(() => {
    getQuestionByIdSpy.mockRestore();
    saveVoteForQuestionSpy.mockRestore();
  });

  it("should render with default QuestionDetails styles", () => {
    const mockData = {
      getQuestionById: jest.fn(),
      saveVoteForQuestion: jest.fn(),
      isLoading: false,
      question: {
        id: "1",
        question: "Who am i ?",
        choices: [
          {
            choice: "a",
            url: "/test/1",
            votes: 0,
            percentage: 0
          },
          {
            choice: "b",
            url: "/test/2",
            votes: 1,
            percentage: 100
          }
        ]
      }
    };
    const { asFragment } = render(<QuestionDetails {...mockData} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with default QuestionDetailsSkelton styles", () => {
    const mockData = {
      getQuestionById: jest.fn(),
      saveVoteForQuestion: jest.fn(),
      isLoading: true,
      question: {
        id: "1",
        question: "Who am i ?",
        choices: [
          {
            choice: "a",
            url: "/test/1",
            votes: 0,
            percentage: 0
          },
          {
            choice: "b",
            url: "/test/2",
            votes: 1,
            percentage: 100
          }
        ]
      }
    };
    const { asFragment } = render(<QuestionDetails {...mockData} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should Invoke getQuestionById and saveVoteForQuestion", async () => {
    const mockData = {
      getQuestionById: getQuestionByIdSpy,
      saveVoteForQuestion: saveVoteForQuestionSpy,
      isLoading: false,
      question: {
        id: "1",
        question: "Who am i ?",
        choices: [
          {
            choice: "a",
            url: "/test/1",
            votes: 0,
            percentage: 0
          },
          {
            choice: "b",
            url: "/test/2",
            votes: 1,
            percentage: 100
          }
        ]
      }
    };

    await waitFor(() => {
      render(<QuestionDetails {...mockData} />);
    });
    expect(getQuestionByIdSpy).toBeCalled();

    const firstSaveVoteBtn = screen.queryByTestId("save-vote-0");
    fireEvent.click(firstSaveVoteBtn);

    expect(saveVoteForQuestionSpy).toBeCalled();
  });
});
