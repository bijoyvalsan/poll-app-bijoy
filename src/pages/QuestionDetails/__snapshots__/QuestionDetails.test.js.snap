// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionDetails Component should render with default QuestionDetails styles 1`] = `
<DocumentFragment>
  <div
    class="sc-bqGGPW bvGpDM"
  >
    <div
      class="sc-ksluID eHYyPg"
    >
      <h2
        class="sc-hBMUJo igLrVS"
        data-testid="page-container-title"
      >
        Question Details
      </h2>
    </div>
    <div>
      <div
        class="sc-dIsUp jCGQWg"
      >
        <p
          class="sc-crzoAE fvbMtY"
        >
          Question: Who am i ?
        </p>
        <div
          class="sc-fujyAs fQbUxZ"
        >
          <div
            class="sc-pNWdM jgNFqv"
          >
            <div
              class="sc-jrsJWt cKvXvH"
            >
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  Choice 1: a
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  Votes: 0
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  Percentage: 0%
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <button
                  class="sc-iCoGMd jtHMFE"
                  data-testid="save-vote-0"
                >
                  Save Vote
                </button>
              </div>
            </div>
          </div>
          <div
            class="sc-pNWdM jgNFqv"
          >
            <div
              class="sc-jrsJWt cKvXvH"
            >
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  Choice 2: b
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  Votes: 1
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  Percentage: 100%
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <button
                  class="sc-iCoGMd jtHMFE"
                  data-testid="save-vote-1"
                >
                  Save Vote
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`QuestionDetails Component should render with default QuestionDetailsSkelton styles 1`] = `
<DocumentFragment>
  <div
    class="sc-bqGGPW bvGpDM"
  >
    <div
      class="sc-ksluID eHYyPg"
    >
      <h2
        class="sc-hBMUJo igLrVS"
        data-testid="page-container-title"
      >
        Question Details
      </h2>
    </div>
    <div>
      <div
        class="sc-dIsUp jCGQWg"
      >
        <div>
          <div
            class="sc-dlnjwi sc-hKFxyN jkEfMo goIlw"
            width="100px"
          />
        </div>
        <div
          class="sc-fujyAs fQbUxZ"
        >
          <div
            class="sc-pNWdM jgNFqv"
          >
            <div
              class="sc-jrsJWt cKvXvH"
            >
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  <div
                    class="sc-dlnjwi sc-hKFxyN hEEJwU goIlw"
                    width="100px"
                  />
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  <div
                    class="sc-dlnjwi sc-hKFxyN hEEJwU goIlw"
                    width="100px"
                  />
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  <div
                    class="sc-dlnjwi sc-hKFxyN hEEJwU goIlw"
                    width="100px"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-pNWdM jgNFqv"
          >
            <div
              class="sc-jrsJWt cKvXvH"
            >
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  <div
                    class="sc-dlnjwi sc-hKFxyN hEEJwU goIlw"
                    width="100px"
                  />
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  <div
                    class="sc-dlnjwi sc-hKFxyN hEEJwU goIlw"
                    width="100px"
                  />
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  <div
                    class="sc-dlnjwi sc-hKFxyN hEEJwU goIlw"
                    width="100px"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-pNWdM jgNFqv"
          >
            <div
              class="sc-jrsJWt cKvXvH"
            >
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  <div
                    class="sc-dlnjwi sc-hKFxyN hEEJwU goIlw"
                    width="100px"
                  />
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  <div
                    class="sc-dlnjwi sc-hKFxyN hEEJwU goIlw"
                    width="100px"
                  />
                </div>
              </div>
              <div
                class="sc-kEqXSa hRaqZx"
              >
                <div
                  class="sc-iqAclL kizGNn"
                >
                  <div
                    class="sc-dlnjwi sc-hKFxyN hEEJwU goIlw"
                    width="100px"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
